README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

hues
====

[![Travis-CI Build
Status](https://travis-ci.org/johnbaums/hues.svg?branch=master)](https://travis-ci.org/johnbaums/hues)
[![CRAN](http://www.r-pkg.org/badges/version/hues)](https://cran.r-project.org/package=hues)

This package generates and plot palettes of optimally distinct colours
in perceptually uniform colour space, based on
[iwanthue](http://tools.medialab.sciences-po.fr/iwanthue/). This is done
through k-means clustering of CIE Lab colour space, according to
user-selected constraints on hue, chroma, and lightness.

`iwanthue` palettes can be added directly onto
[ggplot2](https://github.com/tidyverse/ggplot2) objects via the
functions `scale_colour_iwanthue()` and `scale_fill_iwanthue()`.

### Installation

The current CRAN version of `hues` can be installed with:

``` r
install.packages('hues')
```

The development version of `hues` can be installed using
`install_github` from the
[`devtools`](https://cran.r-project.org/package=devtools) package:

``` r
# install.packages('devtools') # uncomment if devtools isn't already installed
devtools::install_github('johnbaums/hues')
```

### Examples

The examples below show palettes corresponding to some presets given at
the [iwanthue](http://tools.medialab.sciences-po.fr/iwanthue) website.

``` r
library(hues)
iwanthue(5, plot=TRUE)
```

![](man/figures/ex-1.png)

``` r
iwanthue(5, 0, 240, 0, 24, 0, 100, plot=TRUE)    # shades
```

![](man/figures/ex-2.png)

``` r
iwanthue(5, 0, 360, 0, 54, 67, 100, plot=TRUE)   # pastel
```

![](man/figures/ex-3.png)

``` r
iwanthue(5, 0, 360, 54, 180, 27, 67, plot=TRUE)  # pimp
```

![](man/figures/ex-4.png)

``` r
iwanthue(5, 0, 360, 36, 180, 13, 73, plot=TRUE)  # intense
```

![](man/figures/ex-5.png)

``` r
iwanthue(3, 0, 300, 60, 180, 73, 100, plot=TRUE) # fluoro
```

![](man/figures/ex-6.png)

``` r
iwanthue(3, 220, 260, 12, 150, 0, 53, plot=TRUE) # blue ocean
```

![](man/figures/ex-7.png)

### `ggplot2` integration

``` r
library(ggplot2)
ggplot(iris, aes(x=Petal.Width, y=Petal.Length)) +
  geom_point(aes(color=Species), size=3) + 
  scale_colour_iwanthue()
```

![](man/figures/ex2-1.png)

``` r
ggplot(iris, aes(x=Petal.Width, y=Petal.Length)) +
  geom_point(aes(color=Species), size=3) + 
  scale_colour_iwanthue(hmax = 90)
```

![](man/figures/ex2-2.png)

``` r
ggplot(iris, aes(x=Petal.Width, y=Petal.Length)) +
  geom_point(aes(color=Species), size=3) + 
  scale_colour_iwanthue(hmin=0, hmax=300, cmin=60, cmax=180, lmin=73, lmax=100)
```

![](man/figures/ex2-3.png)


NEWS.md

# hues 0.2.0

* `hues` now provides direct support for `ggplot`. The new functions
  `scale_color_iwanthue()` and `scale_fill_iwanthue()` generate a 
  `ScaleDiscrete` object that can be added to a `ggplot` object.

# hues 0.1

* Initial CRAN release.


R/swatch.R
#' Plot colour swatches for a vector of colours
#' 
#' Plot named colour swatches for a vector of colours.
#' 
#' @param x a vector of colours, specified as: colour names (i.e. colour names
#'   returned by [colors()]); numeric indices into [palette()], or hexadecimal
#'   strings in the form `"#RRGGBB"`, where `RR`, `GG`, and `BB` are pairs of
#'   hexadecimal digits representing red, green, and blue components, in the
#'   range `00` to `FF`.
#' @return `NULL`. The colour swatch is plotted to the active plotting device.
#' @seealso [iwanthue]
#' @importFrom graphics barplot par strwidth
#' @export
#' @examples 
#' swatch(colours()[1:10])
#' swatch(1:4)
#' swatch(iwanthue(5))
swatch <- function(x) {
  par(mai=c(0.2, max(strwidth(x, "inch") + 0.4, na.rm = TRUE), 0.2, 0.4))
  barplot(rep(1, length(x)), col=rev(x), space = 0.1, axes=FALSE, 
          names.arg=rev(x), cex.names=0.8, horiz=T, las=1)
  return(invisible(NULL))
}


R/ggplot2_scales.R
#' iwanthue scales to use with ggplot2
#' 
#' These functions allow you to use [iwanthue()] generated palettes with
#' `ggplot2` plots. You need `ggplot2` installed for these to work. Note these
#' only work with discrete scales.
#'
#' @param ... Arguments to pass on to [ggplot2::discrete_scale()].
#' @param aesthetics Character string or vector of character strings listing the
#'   name(s) of the aesthetic(s) that this scale works with. This can be useful,
#'   for example, to apply colour settings to the colour and fill aesthetics at
#'   the same time, via `aesthetics = c("colour", "fill")`.
#' @inheritParams iwanthue
#'
#' @return A `ScaleDiscrete` object that can be added to a `ggplot` object.
#' 
#' @name hues-ggplot2-scales
#' @rdname hues-ggplot2-scales
#' 
#' @author Russell Dinnage <r.dinnage@@gmail.com>
#' @export
#'
#' @examples
#' 
#' if (require('ggplot2')) {
#' 
#'    ggplot(iris, aes(x=Petal.Width, y=Petal.Length)) +
#'      geom_point(aes(color=Species), size=10) + 
#'      scale_colour_iwanthue()
#'     
#'    ggplot(iris, aes(x=Petal.Width, y=Petal.Length)) +
#'      geom_point(aes(color=Species), size=10) + 
#'      scale_colour_iwanthue(hmax = 90)
#'      
#' }
scale_colour_iwanthue <- function(..., hmin=0, hmax=360, cmin=0, cmax=180, 
                                  lmin=0, lmax=100, random=FALSE,
                                  aesthetics = "colour") {
  
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop('ggplot2 is required for this functionality', call. = FALSE)
  }
  
  ggplot2::discrete_scale(aesthetics,
                          "iwanthue",
                          iwanthue_discrete(hmin, hmax, cmin, cmax, lmin, lmax, 
                                            random),
                          ...)
}

#' @rdname hues-ggplot2-scales
#' @export
#' 
scale_color_iwanthue <- scale_colour_iwanthue

#' @rdname hues-ggplot2-scales
#' @export
#' 
scale_fill_iwanthue <- function(..., hmin=0, hmax=360, cmin=0, cmax=180, lmin=0, 
                                lmax=100, random=FALSE, aesthetics = "fill") {
  
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop('ggplot2 is required for this functionality', call. = FALSE)
  }
  
  ggplot2::discrete_scale(aesthetics,
                          "iwanthue",
                          iwanthue_discrete(hmin, hmax, cmin, cmax, lmin, lmax, 
                                            random),
                          ...)
}

iwanthue_discrete <- function(hmin, hmax, cmin, cmax, lmin, lmax, random) {
  function(n) {
    hues::iwanthue(n, hmin = hmin, hmax = hmax, cmin = cmin, cmax = cmax, 
                   lmin = lmin, lmax = lmax, plot = FALSE, random = random)
  }
}


R/iwanthue.R
#' Generate a colour palette by k-means clustering of CIE Lab colour space.
#' 
#' Generate a palette of distinct colours through k-means clustering of CIE Lab 
#' colour space.
#' 
#' @param n Numeric. The number of colours to generate.
#' @param hmin Numeric, in the range `[0, 360]`. The lower limit of the hue
#'   range to be clustered.
#' @param hmax Numeric, in the range `[0, 360]`. The upper limit of the hue
#'   range to be clustered.
#' @param cmin Numeric, in the range `[0, 180]`. The lower limit of the chroma
#'   range to be clustered.
#' @param cmax Numeric, in the range `[0, 180]`. The upper limit of the chroma 
#'   range to be clustered.
#' @param lmin Numeric, in the range `[0, 100]`. The lower limit of the
#'   luminance range to be clustered.
#' @param lmax Numeric, in the range `[0, 100]`. The upper limit of the
#'   luminance range to be clustered.
#' @param plot Logical. Should the colour swatches be plotted (using 
#'   [swatch()])?
#' @param random Logical. If `TRUE`, clustering will be determined by the 
#'   existing RNG state. If `FALSE`, the seed will be set to `1` for clustering,
#'   and on exit, the function will restore the pre-existing RNG state.
#' @return A vector of `n` colours (as hexadecimal strings), representing 
#'   centers of clusters determined through k-means clustering of the CIE Lab 
#'   colour space delimited by `hmin`, `hmax`, `cmin`, `cmax`, `lmin` and 
#'   `lmax`.
#' @details Note that `iwanthue` currently doesn't support `hmin` greater than 
#'   `hmax` (which should be allowed, since hue is circular).
#' @references 
#'   * Examples follow those presented at [iwanthue - colors for data
#'   scientists](http://tools.medialab.sciences-po.fr/iwanthue/)
#'   * [iwanthue on GitHub](https://github.com/medialab/iwanthue)
#' @seealso [swatch]
#' @export
#' @importFrom colorspace LAB hex coords
#' @importFrom stats kmeans
#' @importFrom methods as
#' @examples 
#' iwanthue(5)
#' iwanthue(5, plot=TRUE)
#' iwanthue(5, 0, 240, 0, 24, 0, 100, plot=TRUE)    # shades
#' iwanthue(5, 0, 360, 0, 54, 67, 100, plot=TRUE)   # pastel
#' iwanthue(5, 0, 360, 54, 180, 27, 67, plot=TRUE)  # pimp
#' iwanthue(5, 0, 360, 36, 180, 13, 73, plot=TRUE)  # intense
#' iwanthue(3, 0, 300, 60, 180, 73, 100, plot=TRUE) # fluoro
#' iwanthue(3, 220, 260, 12, 150, 0, 53, plot=TRUE) # blue ocean
iwanthue <- function(n, hmin=0, hmax=360, cmin=0, cmax=180, lmin=0, lmax=100, 
                     plot=FALSE, random=FALSE) {
  stopifnot(hmin >= 0, cmin >= 0, lmin >= 0, 
            hmax <= 360, cmax <= 180, lmax <= 100, 
            hmin <= hmax, cmin <= cmax, lmin <= lmax,
            n > 0)
  if(!random) {
    if (exists(".Random.seed", .GlobalEnv)) {
      old_seed <- .GlobalEnv$.Random.seed
      on.exit(.GlobalEnv$.Random.seed <- old_seed)
    } else {
      on.exit(rm(".Random.seed", envir = .GlobalEnv))
    }
    set.seed(1)
  }
  lab <- colorspace::LAB(as.matrix(
    expand.grid(seq(0, 100, 1), seq(-100, 100, 5), seq(-110, 100, 5))))
  if (any((hmin != 0 || cmin != 0 || lmin != 0 ||
           hmax != 360 || cmax != 180 || lmax != 100))) {
    hcl <- as(lab, 'polarLUV')
    hcl_coords <- colorspace::coords(hcl)
    hcl <- hcl[which(hcl_coords[, 'H'] <= hmax & hcl_coords[, 'H'] >= hmin &
                       hcl_coords[, 'C'] <= cmax & hcl_coords[, 'C'] >= cmin & 
                       hcl_coords[, 'L'] <= lmax & hcl_coords[, 'L'] >= lmin), ]
    lab <- as(hcl, 'LAB')    
  }
  lab <- lab[which(!is.na(colorspace::hex(lab))), ]
  clus <- stats::kmeans(colorspace::coords(lab), n, iter.max=50)
  if (isTRUE(plot)) {
    swatch(colorspace::hex(colorspace::LAB(clus$centers)))
  }
  unname(colorspace::hex(colorspace::LAB(clus$centers)))
}
